rfd/0014-session-2FA.md

```diff
@@ -0,0 +1,474 @@
+```
+
+### enforcement
+
+2FA checks per session can be enforced per-role or globally.
+
+#### per-role
+
+This approach is for operators that want extra protection for some high-value
+resources (like a prod DB VM or k8s cluster) but not others (like a test k8s
+cluster), to reduce the friction for users.
+
+A new field `require_session_2fa` in role `options` specifies whether 2FA is
+required. For example, the below privileged role enforces 2FA per session:
+
+```yaml
+kind: role
+version: v3
+metadata:
+  name: prod-admin
+spec:
+  options:
+    require_session_2fa: true
+
+  allow:
+    logins: [root]
+    node_labels:
+      'environment': 'prod'
+```
+
+Assuming there exists node `A` with label `environment: prod` in the cluster.
+User with role `prod-admin` is required to pass the 2FA check before logging
+into node `A`.
+
+Now, if a user also has the role:
+
+```yaml
+kind: role
+version: v3
+metadata:
+  name: dev
+spec:
+  allow:
+    logins: [root]
+    node_labels:
+      'environment': 'dev'
+```
+
+And there exists node `B` with label `environment: dev` in the cluster.
+Then they _don't_ need the 2FA check before logging into `B`, because role
+`dev` doesn't require it.
+
+Generally, if at least one role that grants access to a resource (SSH node, k8s
+cluster, etc.) sets `require_session_2fa: true`, then 2FA check is required.
+It's required even if another role grants access to the same resource without
+2FA.
+
+#### globally
+
+This approach is for operators that want to enforce 2FA usage org-wide, for all
+sessions.
+
+A new field `require_session_2fa` is available under `auth_service`:
+
+```yaml
+# teleport.yaml
+auth_service:
+  require_session_2fa: true
+```
+
+If this field is set to true, it overrides any values set in roles and always
+requires 2FA checks for all sessions.
+
+### certificate changes
+
+x509 and SSH certificates need 2 new pieces of information encoded:
+
+- is this a short-lived cert issued after 2FA?
+- [constraints](#constraints) for the cert usage
+
+When validating a certificate, the Teleport service will check RBAC to see if
+2FA is required per session. If required, the 2FA flag field must be set in the
+certificate.
+
+#### SSH
+
+SSH certs will encode new data in extensions. New extensions are:
+
+- `issued-with-2fa` - ID of the 2FA token used to issue the cert
+- `client-ip` - IP of the client
+- `session-ttl` - string-encoded `time.Duration`, how long the session is
+  allowed to last after connection start
+
+#### x509
+
+x509 certs will encode new data in the Subject extensions, similar to [the
+other custom fields we
+encode](https://github.com/gravitational/teleport/blob/103465ed5a8e20249275b48ac081ef9517ae5aa7/lib/tlsca/ca.go#L180-L260).
+
+New extensions are:
+
+- `IssuedWith2FA` (OID `1.3.9999.1.8`) - ID of the 2FA token used to issue the
+  cert
+- `ClientIP` (OID `1.3.9999.1.9`) - IP of the client
+- `SessionTTL` (OID `1.3.9999.1.10`) - string-encoded `time.Duration`, how long
```

Comment:
Is it better to use timestamp instead? How does it correlate with built-in session expriation?